app/src/main/java/org/systers/mentorship/viewmodels/RelationViewModel.kt

```diff
@@ -34,82 +30,58 @@ class RelationViewModel : ViewModel() {
+                    MentorshipApplication.getContext().getString(
+                            R.string.error_please_check_internet)
+                }
+                is TimeoutException -> {
+                    MentorshipApplication.getContext().getString(R.string.error_request_timed_out)
+                }
+                is HttpException -> {
+                    CommonUtils.getErrorResponse(throwable).message
+                }
+                else -> {
+                    Log.e(TAG, throwable.localizedMessage)
+                    MentorshipApplication.getContext().getString(
+                            R.string.error_something_went_wrong)
+                }
+            }
+            successfulGet.value = false
+        }
     }
 
     /**
      * Cancels a mentorship relation
      */
-    @SuppressLint("CheckResult")
-    fun cancelMentorshipRelation(relationId: Int) {
-        relationDataManager.cancelRelationship(relationId)
-                .subscribeOn(Schedulers.newThread())
-                .observeOn(AndroidSchedulers.mainThread())
-                .subscribeWith(object : DisposableObserver<CustomResponse>() {
-                    override fun onNext(customResponse: CustomResponse) {
-                        message = customResponse.message
-                        successfulCancel.value = true
-                    }
-
-                    override fun onError(throwable: Throwable) {
-                        when (throwable) {
-                            is IOException -> {
-                                message = MentorshipApplication.getContext()
-                                        .getString(R.string.error_please_check_internet)
-                            }
-                            is TimeoutException -> {
-                                message = MentorshipApplication.getContext()
-                                        .getString(R.string.error_request_timed_out)
-                            }
-                            is HttpException -> {
-                                message = CommonUtils.getErrorResponse(throwable).message
-                            }
-                            else -> {
-                                message = MentorshipApplication.getContext()
-                                        .getString(R.string.error_something_went_wrong)
-                                Log.e(TAG, throwable.localizedMessage)
-                            }
-                        }
-                        successfulCancel.value = false
-                    }
-
-                    override fun onComplete() {
-                    }
-                })
+    fun cancelMentorshipRelation(relationId: Int) = viewModelScope.launch {
+        try {
+            message = relationDataManager.cancelRelationship(relationId).message
+            successfulCancel.value = true
+        } catch (throwable: Throwable) {
+            message = when (throwable) {
+                is IOException -> {
```

Comment:
I have a feeling there is loads copy paste in this exceptions. If yes maybe it would be reasonable to make function exception handler or something like that?

Reply:
Maybe something like
```
function Throwable.handleMentorshipNetworkException(){
 return when (this) {
                is IOException -> {
                    MentorshipApplication.getContext().getString(
                            R.string.error_please_check_internet)
                }
                is TimeoutException -> {
                    MentorshipApplication.getContext().getString(R.string.error_request_timed_out)
                }
                is HttpException -> {
                    CommonUtils.getErrorResponse(throwable).message
                }
                else -> {
                    Log.e(TAG, throwable.localizedMessage)
                    MentorshipApplication.getContext().getString(
                            R.string.error_something_went_wrong)
                }
            }
```

Reply:
Ideally I would add sender so we could identify easier where the code has fail.

Reply:
Yes, error catching is duplicated in almost every Viewmodel many times.
I've found an interesting solution* and am working on it, but this looks very well too.

I'll push commits demonstrating it when I finish an example*

Reply:
What do you meann by adding a sender?

As of now, I just did `throwable.printStackTrace()` in `else` branch in the extension fun you posted above